AsyncDisplayKit/Details/ASCollectionDataController.mm

```diff
@@ -0,0 +1,217 @@
+/* Copyright (c) 2015-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+#import "ASCollectionDataController.h"
+
+#import "ASAssert.h"
+#import "ASMultidimensionalArrayUtils.h"
+#import "ASDisplayNode.h"
+#import "ASDisplayNodeInternal.h"
+#import "ASDataController+Subclasses.h"
+
+//#define LOG(...) NSLog(__VA_ARGS__)
+#define LOG(...)
+
+@interface ASCollectionDataController ()
+
+- (id<ASCollectionDataControllerSource>)collectionDataSource;
+
+@end
+
+@implementation ASCollectionDataController {
+  NSMutableDictionary *_pendingNodes;
+  NSMutableDictionary *_pendingIndexPaths;
+}
+
+- (void)prepareForReloadData
+{
+  _pendingNodes = [NSMutableDictionary dictionary];
+  _pendingIndexPaths = [NSMutableDictionary dictionary];
+
+  [[self supplementaryKinds] enumerateObjectsUsingBlock:^(NSString *kind, NSUInteger idx, BOOL *stop) {
+    LOG(@"Populating elements of kind: %@", kind);
+    NSMutableArray *indexPaths = [NSMutableArray array];
+    NSMutableArray *nodes = [NSMutableArray array];
+    [self _populateSupplementaryNodesOfKind:kind withMutableNodes:nodes mutableIndexPaths:indexPaths];
+    _pendingNodes[kind] = nodes;
+    _pendingIndexPaths[kind] = indexPaths;
+
+    // Measure loaded nodes before leaving the main thread
+    [self layoutLoadedNodes:nodes ofKind:kind atIndexPaths:indexPaths];
+  }];
+}
+
+- (void)willReloadData
+{
+  [_pendingNodes enumerateKeysAndObjectsUsingBlock:^(NSString *kind, NSMutableArray *nodes, BOOL *stop) {
+    // Remove everything that existed before the reload, now that we're ready to insert replacements
+    NSArray *indexPaths = [self indexPathsForEditingNodesOfKind:kind];
+    [self deleteNodesOfKind:kind atIndexPaths:indexPaths completion:nil];
+    
+    NSArray *editingNodes = [self editingNodesOfKind:kind];
+    NSMutableIndexSet *indexSet = [[NSMutableIndexSet alloc] initWithIndexesInRange:NSMakeRange(0, editingNodes.count)];
+    [self deleteSectionsOfKind:kind atIndexSet:indexSet completion:nil];
+
+    // Insert each section
+    NSUInteger sectionCount = [self.collectionDataSource dataController:self numberOfSectionsForSupplementaryNodeOfKind:kind];
+    NSMutableArray *sections = [NSMutableArray arrayWithCapacity:sectionCount];
+    for (int i = 0; i < sectionCount; i++) {
+      [sections addObject:[[NSMutableArray alloc] init]];
```

Comment:
Tiny nit, but might as well be [NSMutableArray array] as the rest of the file uses.